network/p2p/dht_test.go

```diff
@@ -0,0 +1,219 @@
+	require.Eventually(suite.T(), fullyConnectedGraph, time.Second*5, tickForAssertEventually, "nodes failed to discover each other")
+
+	// Step 4: publish a message to the topic
+	require.NoError(suite.T(), dhtServerNode.Publish(suite.ctx, topic, []byte("hello")))
+
+	// Step 5: By now, all peers would have been discovered and the message should have been successfully published
+	// A hash set to keep track of the nodes who received the message
+	recv := make(map[flow.Identifier]bool, count)
+	for i := 0; i < count; i++ {
+		select {
+		case res := <-ch:
+			recv[res] = true
+		case <-time.After(3 * time.Second):
+			var missing flow.IdentifierList
+			for _, n := range dhtClientNodes {
+				if _, found := recv[n.id]; !found {
+					missing = append(missing, n.id)
+				}
+			}
+			assert.Fail(suite.T(), " messages not received by nodes: "+strings.Join(missing.Strings(), ","))
+			break
+		}
+	}
+
+	// Step 6: unsubscribes all nodes from the topic
+	for _, n := range nodes {
+		assert.NoError(suite.T(), n.UnSubscribe(topic))
+	}
+}
+
+// CreateNode creates the given number of libp2pnodes
+// if dhtServer is true, the DHTServer is used as for Discovery else DHTClient
+func (suite *DHTTestSuite) CreateNodes(count int, dhtServer bool) (nodes []*Node) {
+
+	// keeps track of errors on creating a node
+	var err error
+	logger := log.Output(zerolog.ConsoleWriter{Out: os.Stderr}).With().Caller().Logger()
+	defer func() {
+		if err != nil && nodes != nil {
+			// stops all nodes upon an error in starting even one single node
+			suite.StopNodes(nodes)
+		}
+	}()
+
+	handlerFunc := func(network.Stream) {}
+
+	// creating nodes
+	for i := 1; i <= count; i++ {
+		_, key := generateNetworkingAndLibP2PKeys(suite.T())
+		noopMetrics := metrics.NewNoopCollector()
+
+		pingInfoProvider, _, _ := MockPingInfoProvider()
+
+		connManager := NewConnManager(logger, noopMetrics)
+		libP2PHost, err := LibP2PHost(context.Background(), "0.0.0.0:0", key,
+			WithLibP2PConnectionManager(connManager))
+		require.NoError(suite.T(), err)
+
+		var dhtDiscovery *discovery.RoutingDiscovery
+		if dhtServer {
+			dhtDiscovery, err = NewDHTServer(suite.ctx, libP2PHost)
+			require.NoError(suite.T(), err)
+		} else {
+			dhtDiscovery, err = NewDHTClient(suite.ctx, libP2PHost)
+			require.NoError(suite.T(), err)
+		}
+
+		psOption := pubsub.WithDiscovery(dhtDiscovery)
+
+		libp2pPubSub, err := DefaultPubSub(suite.ctx, libP2PHost, psOption)
+		require.NoError(suite.T(), err)
+
+		n, err := NewLibP2PNode(flow.Identifier{}, rootBlockID, logger, libP2PHost, libp2pPubSub,
+			WithConnectionManager(connManager),
+			WithPingService(rootBlockID, pingInfoProvider))
+		require.NoError(suite.T(), err)
+
+		n.SetFlowProtocolStreamHandler(handlerFunc)
+
+		require.Eventuallyf(suite.T(), func() bool {
+			ip, p, err := n.GetIPPort()
+			return err == nil && ip != "" && p != ""
+		}, 3*time.Second, tickForAssertEventually, fmt.Sprintf("could not start node %d", i))
```

Comment:
Do nodes have a `Start` method? or do they immediately start upon construction?

Reply:
it does not have a start method but i investigated this more and the libp2p method `libp2p.New` which is called by our `LibP2PHost` will not return till the listener on the port has been started. So am removing this `Eventually` clause completely.